### PR TITLE
Jakarta Faces 4.1

### PR DESCRIPTION
**Pull request by arjantijms**
## Specification PR template
When creating a specification project release review, create PRs with the content defined as follows.

Include the following in the PR:
- [X] A directory in the form wombat/x.y where x.y is the release major.minor version, and the directory contains the following.
- [X] Specification PDF in the form of jakarta-wombat-spec-x.y.pdf
- [X] Specification HTML in the form of jakarta-wombat-spec-x.y.html
- [X] A specification page named _index.md following the template at:
      https://github.com/jakartaee/specification-committee/blob/master/spec_page_template.md
- [ ] For a Progress Review, that sufficient progress has been made on a Compatible Implementation and TCK, to ensure that the spec is implementable and testable.
- [x] For a [Release Review](https://www.eclipse.org/projects/handbook/#release-review), a summary that a Compatible Implementation is complete, passes the TCK, and that the TCK includes sufficient coverage of the specification. The TCK users guide MUST include the instructions to run the compatible implementations used to validate the release.
Instructions MAY be by reference.
   - [x] Updated release record
   - [x] Email to PMC
   - [x] Start release review by emailing EMO 
- [X] The URL of the OSSRH staging repository for the api, javadoc:
      https://jakarta.oss.sonatype.org/content/repositories/staging/jakarta/faces/jakarta.faces-api/4.1.0/
- [X] The URL of the staging directory on downloads.eclipse.org for the proposed EFTL TCK binary:
      https://download.eclipse.org/faces/jakartaee11/staged/eftl/
- [X] The URL of the compatibility certification request issue:
      https://github.com/jakartaee/faces/issues/1926
- [X] Specification JavaDoc in the wombat/x.y/apidocs directory. 
If desired, an optional second PR can be created to contain just the JavaDoc in the `apidocs` directory.

Note: If any item does not apply, check it and mark N/A below it.

### CODE REVIEW

**Comment by mtdelgadoa** [comment]
Hello all, 
the EMO is trying to avoid duplication of release records, so we will be providing our feedback and comments related to this release directly here.

## EMO REVIEW CHECKLIST 

EDP Review status: ONGOING

### **EMO review checklist**
PMI record: URL

_**EF Specification Process**_      
- [ ] Spec Committee Ballot completed

_**Intellectual Property Management**_      
- [ ] All project code has [copyright and license headers](https://www.eclipse.org/projects/handbook/#ip-copyright-headers) correctly applied. ** EMO will scan the code at their discretion **   
- [ ]  All distributed third-party content has been vetted by the IP Due Diligence process (i.e., IP Log has been approved)   

_**Open Source Rules of Engagement**_
- [ ] PMC approval [on the mailing list](https://github.com/eclipsefdn/emo-team/emo/-/blob/main/%7Blink%20to%20the%20message%7D)

General:
- [ ] Project is operating within the mission and scope defined in its top-level project’s charter   
- [ ]  Project is operating within the bounds of its own scope   
- [ ]  Project is operating in an open and transparent manner       
- [ ]  Overall the project is operating according to the Eclipse Development Process.   

Things to check:
- Communication channels advertised
- Advertised communication channels used
- Committers are responding to questions
- Committers are responding to issues
- Committers are responding to pull/merge/review requests

_**Branding and Trademarks**_
The following applies when the project has a custom website.
To the best of our knowledge:
- [ ]  Project content correctly uses Eclipse Foundation trademarks    
- [ ] Project content (code and documentation) does not violate trademarks owned by other organizations   

Things to check:
- [ ]  Project website uses the [project's formal name](https://www.eclipse.org/projects/handbook/#trademarks-website-name) in first and all prominent references
- [ ]   Project website includes a trademark attribution statement
- [ ] Project website footers contain all [necessary elements](https://www.eclipse.org/projects/handbook/#trademarks-website-footer)

_**Legal Documentation**_
Required files:   
- [ ] License files in all repository roots   
- [ ]  README       
- [ ]  CONTRIBUTING (or equivalent)   

Recommended files:
- [ ] NOTICES or equivalent (you can use the [Legal Documentation generator](https://www.eclipse.org/projects/tools/documentation.php) available under committer tools)   
- [ ]  CODE_OF_CONDUCT   
- [ ]  SECURITY   
 
See examples for [Security file](https://gitlab.eclipse.org/security/best-practices/-/blob/main/templates/SECURITY.md) and [Code of Conduct](https://github.com/eclipse/.github/blob/master/CODE_OF_CONDUCT.md).

Required elements:
- [ ]  ECA is referenced/described   

Recommended elements:
- [ ]  Build instructions provided   
- [ ] [Security policy](https://www.eclipse.org/security/policy.php) is described   
 
_**Metadata (PMI)**_
- [ ] The formal name, e.g. "Eclipse Foo™", is used in the project title   
- [ ] The formal name including appropriate marks (e.g, "™") is used in the first mention in the text of the project description, and scope   
- [ ] The project description starts with a single paragraph that can serve as an executive summary   
- [ ] Source code repository references are up-to-date   
- [ ] Download links and information are up-to-date (see [EF handbook for more information on how-to do this](https://www.eclipse.org/projects/handbook/#resources-downloads))   
- [ ] Communication channels listed in the PMI (i.e. public mailing list, forums, etc.)

**Comment by kazumura** [comment]
# Mentor Spec Review Checklist

1. Spec PR
  - [x] PR uses [template](https://github.com/jakartaee/specifications/blob/master/pull_request_template.md)
  - [x] Directory of form {spec}/x.y
  - [x] PDF of form jakarta-{spec}-spec-x.y.pdf ("-spec" preferred but not required)
  - [x] HTML of form jakarta-{spec}-spec-x.y.html ("-spec" preferred but not required)
  - [x] Index page {spec}/x.y/_index.md following [template](https://github.com/jakartaee/specification-committee/blob/master/spec_page_template.md)
  - [x] Index page {spec}/_index.md following [template](https://github.com/jakartaee/specification-committee/blob/master/spec_index_template.md)
  - [ ] No other files (e.g., no jakarta_ee_logo_schooner_color_stacked_default.png)
  - [ ] Staging repository link of the form https://jakarta.oss.sonatype.org/content/repositories/staging/jakarta/{spec}/jakarta.{spec}-api/x.y.z/
  - [ ] EFTL TCK link of the form http://download.eclipse.org/.../+.zip
  - [ ] Compatibility certification link of the form https://github.com/eclipse-ee4j/{project}/#{issue}
  - [ ] (Optional) Second PR for just apidocs

2. _index.md
  - [ ] Link to project release plan of the form https://projects.eclipse.org/projects/ee4j.{spec}/releases/x.y/plan
  - [ ] Link to spec pdf
  - [ ] Link to spec html
  - [ ] Link to apidocs
  - [ ] Link to final TCK download zip file of the form https://download.eclipse.org/jakartaee/{spec}/x.y/*{spec}-tck-x.y.z.zip (The folder path is required, the file name pattern is preferred.)
  - [ ] Link to API jar file of the form https://search.maven.org/artifact/jakarta.{spec}/jakarta.{spec}-api/x.y.z/jar
  - [ ] Name of and link to at least one Compatible Implementation
  - [ ] Review summary statement to ensure it has been updated to reflect the work accomplished and does not contain proposal statements as might be found in Plan Review statement.

3. javadocs
  - [ ] Footer contains Eclipse copyright and link to license
  - [ ] ESFL license is included, usually as doc-files/speclicense.html
  - [ ] no META-INF directory in PR
  - [ ] javadocs-jar artifact matches apidocs (optional for this release)

4. Spec PDF
  - [ ] Correct spec title
  - [ ] Version number of the form x.y, not x.y.z
  - [ ] Correct Eclipse copyright line
  - [ ] No DRAFT or SNAPSHOT
  - [ ] Correct Logo

5. Spec HTML
  - [ ] Same as PDF

6. TCK zip file
  - [ ] README file (optional for this release)
  - [ ] EFTL license file, preferably named LICENSE.md
  - [ ] User's Guide (or equivalent documentation)
  - [ ] How to test the Compatible Implementation(s) listed in _index.md above with the TCK (may be in UG)

7. TCK User's Guide (or equivalent documentation)
  - [ ] Software requirements listed
  - [ ] Installation and configuration described
  - [ ] How to run tests
  - [ ] Where to file challenges

8. Compatibility certification request
  - [ ] Request follows [template](https://github.com/jakartaee/specification-committee/blob/master/compatibility-certification-request.md)
  - [ ] SHA-256 fingerprint matches staged TCK zip file
  - [ ] Request issue has `certification` label.

9. TCK results summary
  - [ ] Page is hosted by Compatible Implementation project
  - [ ] Includes all information from certification request
  - [ ] Summary includes number of tests passed, failed, errors
  - [ ] SHA-256 fingerprint matches staged TCK zip file on cert request

10. If a Release Review is required, the specification project team contacts the EMO to initiate the review by sending an email to emo@eclipse.org.
    (A Release Review is not required if the current release is a Service Release based on a previously successful Major or Minor
    release as indicated by a release record on the project's Releases page, e.g., the [Jakarta Servlet releases page](https://projects.eclipse.org/projects/ee4j.servlet/reviews).)
    - [ ] The specification project team requests approval by sending an email to the EMO (with cc to the PMC) that contains a link to **this** PR and a request to the PMC for approval.
          
11. Update Jakarta EE API jar
  - [ ] Update the Jakarta EE API jar by submitting a PR to the jakartaee-api project that updates the version number of your API jar file.

**Comment by arjantijms** [comment]
@kazumura CCR has been added

**Comment by kazumura** [comment]
@arjantijms   
Thanks for PR.

The followings are missing in this PR. Please add them.
* Jsdoc
* Renderkitdoc
* VDLDoc

**Comment by arjantijms** [comment]
> The followings are missing in this PR. Please add them.

Thanks, it is done!

**Comment by pnicolucci** [comment]
Hi, the `_index.md` should be updated to reflect the changes for Faces 4.1.

We should call out the changes from the [Faces 4.1 Changelog](https://github.com/jakartaee/faces/blob/4.1/spec/src/main/asciidoc/ChangeLog.adoc#changes-between-41-and-40).

These are Faces 5.0 changes that should be removed:
1. https://github.com/jakartaee/faces/issues/206 (Make SelectItem#value generic Component: Components/Renderers) 
2. https://github.com/jakartaee/faces/issues/1564 looks to be mentioned: Re-add #{request} in CDI mode 
3. https://github.com/jakartaee/faces/issues/1564 looks to be mentioned: Re-add #{request} in CDI mode 

Missing Deprecation:

1. Missing Full State Saving Deprecation in the deprecations list: https://github.com/jakartaee/faces/issues/1829

Other:

1. I don't think there is a need to call out: Remove unused PreDestroyCustomScopeEvent and PostConstructCustomScopeEventAdd in both the new features and removals.

**Comment by arjantijms** [comment]
> Hi, the _index.md should be updated to reflect the changes for Faces 4.1.

Thanks, can you verify the updated index?

**Comment by pnicolucci** [comment]
Hi @arjantijms 

We should remove the following as they point to Faces 5.0 work:

1. "importConstants should be allowed everywhere, not only in  f:metadata" as that is in Faces 5.0: https://github.com/jakartaee/faces/issues/1466
2. "Enhance UIInput events with HTML5 like oninput" as that is also Faces 5.0: https://github.com/jakartaee/faces/issues/1507
4. Setting/overriding components default value -> https://github.com/jakartaee/faces/issues/1602

Also I think removing "<f:ajax> execute="@this" and render="@this" does not behave as expected when nested in composite component" would be good since that looks like it was for Faces 4.0: https://github.com/jakartaee/faces/issues/1567

Would be good to add:

Add UUIDConverter -> https://github.com/jakartaee/faces/issues/1819
Add ExternalContext.setResponseContentLengthLong ->  https://github.com/jakartaee/faces/issues/1764
Add rowStatePreserved property to UIRepeat, exactly the same as UIData -> https://github.com/jakartaee/faces/issues/1263
Spec: jakarta.faces.FACELETS_REFRESH_PERIOD default when ProjectStage is Development -> https://github.com/jakartaee/faces/issues/1821
FacesMessage: implement equals(), hashcode(), toString()  -> https://github.com/jakartaee/faces/issues/1823

File: faces/4.1/_index.md [hunk 1]
```diff
@@ -33,21 +32,19 @@ The removes references to the SecurityManager, further aligns with CDI where pos
...
-<!--
-The following can be uncommented and version information updated as they become available.
-
 * [Jakarta Faces 4.1 Specification Document](./jakarta-faces-4.1.pdf) (PDF)
 * [Jakarta Faces 4.1 Specification Document](./jakarta-faces-4.1.html) (HTML)
 * [Jakarta Faces 4.1 Javadoc](./apidocs)
 * [Jakarta Faces 4.1 Jsdoc](./jsdoc)
 * [Jakarta Faces 4.1 Renderkitdoc](./renderkitdoc)
 * [Jakarta Faces 4.1 VDLDoc](./vdldoc)
 * [Jakarta Faces 4.1 TCK](https://download.eclipse.org/jakartaee/faces/4.1/jakarta-faces-tck-4.1.0.zip)([sig](https://download.eclipse.org/jakartaee/faces/4.1/jakarta-faces-tck-4.1.0.zip.sig),[sha](https://download.eclipse.org/jakartaee/faces/4.1/jakarta-faces-tck-4.1.0.zip.sha256),[pub](https://raw.githubusercontent.com/jakartaee/specification-committee/master/jakartaee-spec-committee.pub))
```
Comment (Pandrex247) [inline faces/4.1/_index.md:41]: Spaces for readability
```suggestion
* [Jakarta Faces 4.1 TCK](https://download.eclipse.org/jakartaee/faces/4.1/jakarta-faces-tck-4.1.0.zip) ([sig](https://download.eclipse.org/jakartaee/faces/4.1/jakarta-faces-tck-4.1.0.zip.sig), [sha](https://download.eclipse.org/jakartaee/faces/4.1/jakarta-faces-tck-4.1.0.zip.sha256), [pub](https://raw.githubusercontent.com/jakartaee/specification-committee/master/jakartaee-spec-committee.pub))
```

File: faces/4.1/_index.md [hunk 2]
```diff
@@ -33,21 +32,19 @@ The removes references to the SecurityManager, further aligns with CDI where pos
 
 * [Jakarta Faces 4.1 Release Record](https://projects.eclipse.org/projects/ee4j.faces/releases/4.1)
```
Comment (Pandrex247) [inline faces/4.1/_index.md:33]: Most other specs also include a link to the Platform release plan:
```suggestion
* [Jakarta Faces 4.1 Release Record](https://projects.eclipse.org/projects/ee4j.faces/releases/4.1)
  * [Jakarta EE 11 Platform Release Plan](https://jakartaee.github.io/platform/jakartaee11/JakartaEE11ReleasePlan)
```

**Review by Pandrex247 — COMMENTED** [review]
Just a couple of things

File: faces/4.1/_index.md (hunk 2 above)
Comment (arjantijms) [reply faces/4.1/_index.md:33]: Looks good, thanks Andy! :)

**Comment by kazumura** [comment]
@arjantijms   
Please look into @pnicolucci comment at "https://github.com/jakartaee/specifications/pull/738#issuecomment-2123248681" .

**Comment by arjantijms** [comment]
> Please look into @pnicolucci comment at "https://github.com/jakartaee/specifications/pull/738#issuecomment-2123248681" .

I just did, thanks! @pnicolucci can you do the final check?

**Comment by pnicolucci** [comment]
@arjantijms  this looks good! My only minor comment is that the UUID entry points to an issue but nothing else does. So for consistency, we could remove the issue reference, or add an issue for each item.

**Comment by kazumura** [comment]
On ballot completion, the specification committee mentor:

 - [x] adds this final checklist to the main PR.
 - [x] adds the `approved` label to the PRs, and sends out the Ballot Summary per this [template](#ballot-summary-email-template) to the [public Jakarta EE Specification Committee email list](jakarta.ee-spec@eclipse.org)
 - [ ] calculates the staged EFTL TCK signature and promotes it to the committee download area
  using the https://ci.eclipse.org/jakartaee-spec-committee/job/promote-release/ job. Manually editing the jenkins Build Information will help identify the build (ie. Mail 2.0 or CDI 3.0).
 - [ ] merges the specification (and apidocs) PRs, ensuring the "date:" field in the _index.md file has an appropriate value to allow publishing.
 - [ ] updates the specification page with the ballot results. This is normally done via a separate PR that should be reviewed, approved, and merged.
 - [x] notifies the EMO of the ballot results by email to [emo@eclipse-foundation.org](emo@eclipse-foundation.org). Just forward the ballot summary note sent earlier to the public Spec Committee email list.
 - [ ] creates an issue in the specification project that includes the following checklist for the specification project team:

**Pull request merged**